### PR TITLE
Refactored code, added new Button class, reduced rerenders

### PR DESCRIPTION
**Pull request by jriyyya**
#23 stated a lot of software design issues and issues with performance and also the architecture of the codebase
while this PR does not aim to be a direct resolution to that issue, this PR does fix a lot of problems mentioned in it

### CODE REVIEW

**Comment by jriyyya** [comment]
There's an issue I'm facing and that is the fact that in the scorewindow after a game is over when I press the `back home` button it does not take me back home, it does sometimes but it is unreliable and might sometimes send me to another gamemode.
I tried but was not able to find a solution to this. 

@chimosky  @quozl  would you please help me with this?

**Comment by quozl** [comment]
Hello.  I'm on leave right now, and was able to take a quick look.

* try to stick to 80 column limit for git commit messages; your messages were truncated in gitk, and badly wrapped when using Terminal to view, see our guide to contributing for suggested layout,
* the code starts out being hard to read, even before your changes, with hundreds of flake8 excursions; feel free to run autopep8 and commit the changes with a commit message that says autopep8 was run,
* our review standards are way lower for activities that don't yet work, or haven't been published, so we might not be as picky and careful as you see us in other repositories,
* sugargame 1.3 is used, but the activity does _not_ follow the sample activity exactly; for instance it calls show_all in the toolbar builder, which implies it has been an unfinished port to 1.3,
* the way that different event loops are used makes it hard to figure out what is actually going on when the problem occurs; you may need to use _print_ statements or an interactive debugger to find out which loop is active at any given time.

In testing [4d6dc05](https://github.com/sugarlabs/make-them-fall-activity/pull/25/commits/4d6dc050d121ef32f4d1c69bf204ffabbbbd1e50), I reproduced what you described, and also an exception;

```log
Traceback (most recent call last):
  File "/usr/share/sugar/activities/MakeThemFall.activity/main.py", line 165, in run
    btn.update()
  File "/usr/share/sugar/activities/MakeThemFall.activity/button.py", line 39, in update
    self.action()
  File "/usr/share/sugar/activities/MakeThemFall.activity/main.py", line 141, in <lambda>
    self.buttons.append(Button(390, 450, "data/images/welcomescreen/2paneheart.png", lambda : self.run_game(pane2heartwindow, 6), maxcardiac))
  File "/usr/share/sugar/activities/MakeThemFall.activity/main.py", line 75, in run_game
    if scorewindow(self.gameDisplay, s, gamenumber).run():
  File "/usr/share/sugar/activities/MakeThemFall.activity/scorescreen.py", line 59, in __init__
    if score > maxscore[gamenumber - 1]:
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

The exception was not reproducible in repeated execution unless score.pkl was removed from the activity root directory.

Hope that helps!

**Comment by quozl** [comment]
Interesting, thanks.  4d6dc05 was 326 flake8 messages.  17bdd8b brings it down to 240.  A further `--experimental --aggressive --aggressive` brings it down to 153.

I've looked at the problem you described.  An unusual excursion from the Sugargame test activity is that Make Them Fall calls `game.start()` in `Activity.__init__`.  This is wrong to begin with, as it causes Pygame initialisation at the wrong time relative to GTK initialisation.  This shows incomplete port to Sugargame 1.2, a520472d9d1aa6baf203868b41cb1684b579f015.

Later, `game.run()` renders the buttons, and the buttons call `game.run_game()` which calls the event loop in the other source files.  After that event loop finishes, it calls `scorewindow`, and if the _Back Home_ button is pressed `scorewindow.run()` is to return `0`, and `run_game()` runs `start()` again.  This makes no sense to me.

It is not a pattern I've seen in any of the other activities, so it was a pattern introduced by this activity.  (Other activities that try to make a button class in Pygame are Pycut and Math Hurdler).

Further debugging will be required. :grin:

**Comment by quozl** [comment]
Okay, I've found the problem with some `print()` statements;

```diff
--- a/main.py
+++ b/main.py
@@ -56,6 +55,7 @@ class game:
     sound = True
 
     def __init__(self):
+        print("game.__init__")
         self.crashed = False
         self.running_mode = None
         self.clock = pygame.time.Clock()
@@ -69,6 +69,7 @@ class game:
         self.maxscore = [0, 0, 0, 0, 0, 0]
 
     def run_game(self, window, gamenumber):
+        print("game.run_game entry", window, gamenumber)
         self.running_mode = window()
         self.running_mode.crashed = self.crashed
         s = self.running_mode.run(self.gameDisplay, self.info)
@@ -77,6 +78,7 @@ class game:
             self.run_game(window, gamenumber)
 
         self.start()
+        print("game.run_game return", window, gamenumber)
 
     def show_help(self):
         a = rules()
@@ -92,6 +94,7 @@ class game:
                 self.maxscore = pickle.load(inp)
 
     def start(self):
+        print("game.start")
 
         self.gameDisplay = pygame.display.get_surface()
 
@@ -155,9 +200,14 @@ class game:
 
     def run(self):
+        print("game.run entry")
         self.gameDisplay = pygame.display.get_surface()
 
         while not self.crashed:
@@ -169,6 +219,7 @@ class game:
 
             event = pygame.event.poll()
             if event.type == pygame.QUIT:
+                print("game.run quit")
                 return
 
             for btn in self.buttons:
@@ -178,6 +229,7 @@ class game:
             pygame.display.update()
             self.clock.tick(60)
 
+        print("game.run return")
         return
 

```

Briefly, the welcome screen buttons think they have been pressed.

In turn, because the score window button has just been pressed and the mouse is still down.

You can see this by pressing the _Back Home_ button in the lower right corner, which corresponds to no button on the welcome screen.  The welcome screen is shown.

If you click on the top right of the _Back Home_ button, the welcome screen detects this as _Cardiac_ game, and begins to play it.  Here's the log;

```log
game.__init__
game.start
game.run entry
game.run_game entry <class 'normal.pane2window'> 1
game.start
game.run_game return <class 'normal.pane2window'> 1
game.run_game entry <class 'normal.pane2window'> 1
game.start
game.run_game return <class 'normal.pane2window'> 1
game.run_game entry <class 'normal.pane2window'> 1
game.start
game.run_game return <class 'normal.pane2window'> 1
game.run_game entry <class 'normal.pane2window'> 1
game.start
game.run_game return <class 'normal.pane2window'> 1
game.run_game entry <class 'cardiac.pane2heartwindow'> 6
game.start
game.run_game return <class 'cardiac.pane2heartwindow'> 6
```

A couple of ways you could fix this;
* wait for mouse button or key release before returning from a window,
* change the button class to work like **almost every other GUI on the planet**; which take action when a mouse button is _released_ not when it is _pressed_.

**Comment by jriyyya** [comment]
I have made some changes, I will be making a few commits for flake8 fixes.

File: main.py
Comment (chimosky) [inline main.py:51]: Makes sense for these to be constants as they're globals.

File: main.py
Comment (chimosky) [inline main.py:42]: It'll be great to get rid of these imports and use relative/absolute imports.

File: main.py
Comment (chimosky) [inline main.py:76]: How come you're passing args to `scorewindow` when `scorewindow.run` requires the args and not `scorewindow`?

File: main.py
Comment (chimosky) [inline main.py:83]: Why create `a` and then pass it to `self.running_mode` and then reassign a later to the value of `a.run()`?
Using variable names in this context would be great, `_rules` instead of `a` as it's used only within the function.

File: main.py
Comment (chimosky) [inline main.py]: You moved this functionality to `update_highscore` so why not use that?

File: main.py
Comment (chimosky) [reply main.py:76]: Seen your refactor.

File: main.py
Comment (jriyyya) [reply main.py:42]: Yes, you can leave the flake8 warnings, I have covered most of them in the next commit, pushing them soon

File: cardiac.py
Comment (chimosky) [inline cardiac.py:8]: The import of `gi` and use of `gi.require_version` should come before any other imports from `gi`.

File: howtoplay.py
Comment (chimosky) [inline howtoplay.py]: Also remove the no new line at EOF.

File: impossible.py
Comment (chimosky) [inline impossible.py:8]: This too.

File: impossible.py
Comment (chimosky) [inline impossible.py]: This too.

File: inferno.py
Comment (chimosky) [inline inferno.py:8]: This too.

File: inferno.py
Comment (chimosky) [inline inferno.py]: This too.

File: main.py
Comment (chimosky) [inline main.py]: This and every other like it.

File: nightmare.py
Comment (chimosky) [inline nightmare.py:8]: This and every other like it.

File: scorescreen.py
Comment (chimosky) [inline scorescreen.py]: Typo here.

**Review by chimosky — CHANGES_REQUESTED** [review]

**Comment by chimosky** [comment]
Reviewed 31af7c4, not tested.

**Comment by quozl** [comment]
Please don't put in `gi.require_version` calls unless they have an effect.  Usually one is enough, in the main program called by the Sugar shell.  Redundant calls waste time, both execution time and time taken to understand the code.

**Comment by jriyyya** [comment]
I see, It was done by the original author I suppose, I haven't taken a much look in other files, except for the main.py and scorescreen.py which I refactored, 

I just did auto pep8 in all files because of which it was in my commits, Fixing it now though 🫡

**Comment by chimosky** [comment]
Tested 1d2f127, no new issues were found.

I'll wait for @quozl to review before I merge.

**Comment by chimosky** [comment]
If 861d2c9 would lead to other flake8 fixes, it'll be better to put it in a separate PR else it's fine as is.

**Comment by quozl** [comment]
Sorry for the delay.  Reviewed the aggregate change (not the commits), and have following comments;

* the activity `__init__` builds the toolbar after creating the PygameCanvas object, which is different to how the Sugargame test activity shows how to do it; if there's not a good reason to depart from this model, it would be best to have it the same, and if there is a good reason it should be in comments,
* the activity `__init__` calls game.start; instead `game.start` can be called from the end of `game.__init__`, again so that the code complies with the model implementation.

**Comment by chimosky** [comment]
>     * the activity `__init__` builds the toolbar after creating the PygameCanvas object, which is different to how the Sugargame test activity shows how to do it; if there's not a good reason to depart from this model, it would be best to have it the same, and if there is a good reason it should be in comments,
> 
>     * the activity `__init__` calls game.start; instead `game.start` can be called from the end of `game.__init__`, again so that the code complies with the model implementation.


Agreed, thanks for pointing them out.

**Comment by jriyyya** [comment]
The only reason the order for both the calls `game.start` and `build_toobar` is in this particular manner
is because if the display is not initialized (activity.py line 26) and set to the game (line 30). we can't run game.start
As it would lead to an error that video is unitialized
Similarly moving the build_toolbar results in the activity crashing without any errors in the console or logs.

**Comment by chimosky** [comment]
> The only reason the order for both the calls `game.start` and `build_toobar` is in this particular manner is because if the display is not initialized (activity.py line 26) and set to the game (line 30). we can't run game.start As it would lead to an error that video is unitialized Similarly moving the build_toolbar results in the activity crashing without any errors in the console or logs.

Could you share your logs?

The display is initialized when you assign `self. game.canvas` as the modules to be initialized are passed and sugargame initializes them.

**Comment by jriyyya** [comment]
Here are the logs

```
Traceback (most recent call last):
File "/usr/bin/sugar-activity3", line 5, in <module>
activityinstance.main()
File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 230, in main instance = create_activity_instance (activity_constructor, activity_handle)
File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 59, in create_activity_instance activity constructor (handle)
File "/home/user/fall/activity.py", line 25, in __init__
self.game = main.game()
File "/home/user/fall/main.py", line 68, in __init_
self.start()
File "/home/user/fall/main.py", line 96, in start
self.info pygame.display. Info() pygame.error: video system not initialized
```

**Comment by chimosky** [comment]
This should fix your issue;

```python3
--- a/activity.py
+++ b/activity.py
@@ -23,14 +23,16 @@ class Activity(activity.Activity):
         self.max_participants = 1
         self.sound = True
         self.game = main.game()
+
+        self.build_toolbar()
+
         self.game.canvas = sugargame.canvas.PygameCanvas(
                 self,
-                main=self.game.run,
+                main=self.game.start,
                 modules=[pygame.display, pygame.font, pygame.mixer])
         self.set_canvas(self.game.canvas)
         self.game.canvas.grab_focus()
-        self.build_toolbar()
-        self.game.start()
+        #self.game.start()
 
     def build_toolbar(self):
 
diff --git a/main.py b/main.py
index 0a07712..999aab1 100644
--- a/main.py
+++ b/main.py
@@ -166,28 +166,29 @@ class game:
         howto = pygame.image.load("data/images/welcomescreen/howtoplay.png")
         self.gameDisplay.blit(howto, (490, 550))
 
-    def run(self):
-        self.gameDisplay = pygame.display.get_surface()
+    #def run(self):
+        #self.gameDisplay = pygame.display.get_surface()
+        #self.start()
+
+        for button in self.buttons:
+            button.update()
 
         while not self.crashed:
             # Gtk events
             while Gtk.events_pending():
                 Gtk.main_iteration()
-            if self.crashed:
-                break
-
-            event = pygame.event.poll()
-            if event.type == pygame.QUIT:
-                return
+            #if self.crashed:
+                #break
 
-            for btn in self.buttons:
-                btn.update()
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    self.crashed = True
```

This change starts the game but hangs at the startup screen, you might have to rename `start` to `run` if you combine them.

I didn't have much time to look indepth into what's causing the activity not to run as expected, I'll leave the debugging to you and you can use lots of print statements to help you along the way.

**Comment by jriyyya** [comment]
We can remove the `game.start` from activity.py if we move it over to the `run()` function. that's fine if our goal is to remove it from activity.py but if we want to add it to `__init__` that would take in depth changes as we are initializing the game before we assign to it a PygameCanvas. 
Also your changes are not working for me, the activity simply won't start and I tried working around your changes a little to no avail. 

finally, build_toolbar again, because we need to set the canvas to the pygame game before we call buid_toolbar so what comments should I add to specify this?

**Comment by chimosky** [comment]
> We can remove the `game.start` from activity.py if we move it over to the `run()` function. that's fine if our goal is to remove it from activity.py but if we want to add it to `__init__` that would take in depth changes as we are initializing the game before we assign to it a PygameCanvas. 

The goal is to keep the model implementation as shown in the sugargame test activity  and you can change it for sure but also include your reason as to doing so.

> Also your changes are not working for me, the activity simply won't start and I tried working around your changes a little to no avail.

I stated in my comment that it freezes when the activity starts without showing the pygame canvas and the diff I shared is a fix for the error shown in the log you shared.


> finally, build_toolbar again, because we need to set the canvas to the pygame game before we call buid_toolbar so what comments should I add to specify this?

Why do you need to set the pygame canvas before calling `build_toolbar`?

**Comment by quozl** [comment]
Reviewed again.  Added my changes on top of your branch, be sure to pull from GitHub.  The interesting things I learned were;

* it is important to change `activity.info` at the same time as changing the activity class name (825d3bd199888bc8a7d0d5a13cc2c9418c9f1bbc vs bc1a550304127ab0df28353758ef0fdd6c736333), otherwise you get a blank grey display,
* being able to exclude Sugar was fairly useful (926b14e66de1aa79af53d3044e5431447a992501) because it showed a difference in handling VIDEORESIZE events between use with Sugar and use without,
* the application did not properly redraw the menu background after a resize event, so I didn't transfer that from the test activity.

Please review my changes.
* 
*

**Comment by jriyyya** [comment]
Reviewed your changes, everything looks good to me.

Also I noticed the screen was not centered so I have fixed it, Please review.

File: button.py
Comment (chimosky) [inline button.py:12]: Could you compare with floor division and see how that affects the dimensions?

Also look at comparing other int divisions too.

**Comment by quozl** [comment]
Blank grey display may happen if Sugar is not restarted?  The activity.info file was changed.

**Comment by jriyyya** [comment]
> * only the main menu and help screens are centered, the game modes are not

Okay, will look into it.

**Comment by chimosky** [comment]
Haven't reviewed your latest change yet but I'll say this, split the latest commit into various commits, each change a particular part of the commit.

If you want to give credit to someone then include it in your commit message as that's part of the commit history and kept but GH PRs aren't.

**Comment by jriyyya** [comment]
> Haven't reviewed your latest change yet but I'll say this, split the latest commit into various commits, each change a particular part of the commit.


i have already changd the files a lot and it would be quite difficult for me to phrase each change in a commit, I do not know how I can commit specific changes in a file. Since the code is so different now how would I go ahout creating transitional commits.
what's your opinion

**Comment by chimosky** [comment]
> > Haven't reviewed your latest change yet but I'll say this, split the latest commit into various commits, each change a particular part of the commit.
> 
> i have already changd the files a lot and it would be quite difficult for me to phrase each change in a commit, I do not know how I can commit specific changes in a file. Since the code is so different now how would I go ahout creating transitional commits. what's your opinion

You'll need to first reset the commit and then you can add patches interactively with `git add -i`, this would enable you to select various diffs to be staged and then you can commit those.

**Comment by jriyyya** [comment]
okay got it, I will do so

**Comment by jriyyya** [comment]
@chimosky Please review this as well

File: game.py
Comment (chimosky) [inline game.py:145]: If this is a generic game class, then it shouldn't implement run as any new game mode should implement it's run function where it'll handle all pygame events as this just handles one.

If they're parts of this function that's generic then let those be left and this class can be inherited and any function redefinition can happen if necessary.

File: game.py
Comment (chimosky) [inline game.py:3]: This gets imported in this commit but then the file gets added in a different commit, either move the addition here or move the usage to the addition.

File: spike.py
Comment (chimosky) [inline spike.py:1]: Please add a license here too.

File: spike.py
Comment (chimosky) [inline spike.py:8]: No need for the parentheses here.

File: spike.py
Comment (chimosky) [inline spike.py:14]: Could you move these close to the function definition as it makes it more readable.

File: spike.py
Comment (chimosky) [inline spike.py:15]: No need for the parentheses.

File: spike.py
Comment (chimosky) [inline spike.py:27]: This could be changed to `if self.y < (self.initial - self.path_length):`

File: generator.py
Comment (chimosky) [inline generator.py:25]: I'm curious, why the `guys` param?

**Review by chimosky — CHANGES_REQUESTED** [review]
I'm curious, wouldn't it be better to have a generic name for the file instead of guy.py and a generic name for the class too.

**Review by chimosky — COMMENTED** [review]
While I agree with the creation of a generic Game class, these should also exist as they're supposed to handle different pygame events in different ways. 

They can inherit from the generic Game class and implement their `run` function to handle what needs to be handled, this also makes debugging easier as the game modes are modular.

**Comment by chimosky** [comment]
Reviewed 1a8c7e0, not tested.

Also add licenses to the necessary files.

File: generator.py
Comment (jriyyya) [reply generator.py:25]: checks collision with each guys and returns True if game over, false otherwise

File: game.py
Comment (jriyyya) [reply game.py]: The class is not supposed to inherit from, when I say it's generic I mean that it acts as all the game types itself depending upon the keymap input which it takes.

Keymap also helps the class to handle the event function accordingly in a modular manner thus eliminating a reason for any inhertied classes to implement their own run function or even inherit from it in the first place.

Comment (chimosky) [reply game.py]: Keymap doesn't help the class handle events accordingly as keymaps is a list and event handling from your changes are happening within the `pygame.event.get()` loop which at the time only handled one event.

File: game.py
Comment (chimosky) [inline game.py:40]: How come this has two loops when the old code had just one?

File: game.py
Comment (chimosky) [reply game.py:22]: You could squash both commits as that's easier.

Adding your personal contact details to the license isn't a great idea, please remove it.

Comment (jriyyya) [reply game.py:22]: Can you please tell me how can I squash ( [19986b6](https://github.com/sugarlabs/make-them-fall-activity/pull/25/commits/19986b60de562997dcf5821c806ff3bedf27efdd) and [44c92ca](https://github.com/sugarlabs/make-them-fall-activity/pull/25/commits/44c92cab94c42b5b31f171f6420e1d5a92a2259f) ) these two commits, I tried but it is getting real messy

File: game.py
Comment (jriyyya) [reply game.py]: I've made a commit related to this [0b6a605](https://github.com/sugarlabs/make-them-fall-activity/pull/25/commits/0b6a605c6ecef4bc87a1582e08c5375c2fa8c4d3)

File: game.py
Comment (jriyyya) [reply game.py:37]: I'm using list comprehension for flattening the 2-D array which was taken as an input in the constructor

File: game.py
Comment (chimosky) [reply game.py:19]: You can run an interactive rebase on the parent commit and in this case that's 363793c so it'll look something like this;
```bash
# This should open your editor and then you can look for both commits and move them close to each other
$ git rebase -i 363793c
# So in your editor you'll have something like this
pick 19986b6 Created generic Game class
pick efddd14 Create class Spike in spike.py
pick 44c92ca created Generator class in generator.py
# Which you'll rearrange to
pick 19986b6 Created generic Game class
s 44c92ca created Generator class in generator.py
pick efddd14 Create class Spike in spike.py
# Then exit the editor and you'll be taken into another editor to edit both commit messages and you can 
# then combine them into one and delete the other so you'll have one commit message and the 
# other would be empty then exit
$ git rebase --continue # should be the final trick.

File: generator.py
Comment (chimosky) [reply generator.py:41]: Yes, I should have phrased that differently.

I meant we should change the variable to something more generic like sprite.

File: game.py
Comment (chimosky) [reply game.py:37]: Yes, basically in the loop, row was added while looping through the keymap but now you're also looping through the row and adding each item in row when in your loop you just added the rows.

Why change to doing it that way?

File: game.py
Comment (chimosky) [reply game.py]: Thanks!

**Comment by chimosky** [comment]
Reviewed 0b6a605, not tested.

File: game.py
Comment (jriyyya) [reply game.py:19]: Thanks Ibiam, I should have asked beforehand, I spent a lot of time figuring this out while completely forgetting that I can change the order of the commits in the rebase 😅

File: generator.py
Comment (jriyyya) [reply generator.py:41]: So is Sprite name fine? Should I change this everywhere guy name is used

File: game.py
Comment (chimosky) [reply game.py:37]: Okay.

File: generator.py
Comment (chimosky) [reply generator.py:41]: Yes sprite is fine if you can't think of anything else, turtleart uses sprites that's why it came to my mind.

**Comment by chimosky** [comment]
Tested 77ada49, key presses aren't handled and this is seen in the logs.

```python3
Traceback (most recent call last):
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 128, in _keydown_cb
    return self._keyevent(widget, event, pygame.KEYDOWN)
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 169, in _keyevent
    mod = self._keymods()
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 145, in _keymods
    mod |= self.__keystate[key_val] and mod_val
IndexError: list index out of range
Traceback (most recent call last):
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 128, in _keydown_cb
    return self._keyevent(widget, event, pygame.KEYDOWN)
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 169, in _keyevent
    mod = self._keymods()
  File "/home/ibiam/Activities/make-them-fall-activity/sugargame/event.py", line 145, in _keymods
    mod |= self.__keystate[key_val] and mod_val
IndexError: list index out of range
```

**Comment by chimosky** [comment]
This PR seems to have achieved it's aim of refactoring the code and #27 was opened to track the poor event handling, @quozl do you have anything to add before I merge?

**Comment by quozl** [comment]
No, nothing springs to mind.

**Pull request merged**
